### PR TITLE
First pass of the Tracker View Model

### PR DESCRIPTION
**Pull request by StevenFrost**
Initial implementation of the Tracker View Model. Added a base class as well since all view models require the view (JavaScript) event dispatcher.

This closes the loop for the `SimCache.TrackerLoadedEvent` event that the tracker view is already sending to the SimCache module.

### CODE REVIEW

File: Build/Config/Cpp.props [hunk 1]
```diff
@@ -38,6 +38,8 @@
       <Optimization Condition="'$(Configuration)' != 'Release'">Disabled</Optimization>
       <BasicRuntimeChecks Condition="'$(Configuration)' == 'Release'"></BasicRuntimeChecks>
 
+      <AdditionalIncludeDirectories>$(SourceDirectory)$(ProjectName)\Source;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (StevenFrost) [inline Build/Config/Cpp.props:41]: This means we don't have relative include paths all over the place when including headers in the source folder.

`#include "../ViewModel.h"` vs `#include "ViewModels/ViewModel.h"`

File: Source/SimCacheModule/Source/ViewModels/TrackerViewModel.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	if ( OnTrackerLoadedEventHandle.IsValid() )
+	{
+		GetViewEventDispatcher().UnregisterEventListener( OnTrackerLoadedEventHandle );
+	}
+}
+
+// -----------------------------------------------------------------------------
+
+void TrackerViewModel::OnTrackerLoaded( const TrackerLoadedEvent& Event )
+{
```
Comment (StevenFrost) [inline Source/SimCacheModule/Source/ViewModels/TrackerViewModel.cpp:67]: To be completed in a later PR when we have the required dependencies.

File: Source/SimCacheModule/Source/ViewModels/ViewModel.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	virtual bool Initialize();
+	virtual void Uninitialize();
+
+protected:
+
+	Utils::EventDispatcher& GetViewEventDispatcher() const;
+
+private:
+
+	Utils::EventDispatcher& ViewEventDispatcher;
```
Comment (StevenFrost) [inline Source/SimCacheModule/Source/ViewModels/ViewModel.h:26]: Intentionally a ref this time! The event dispatcher is owned by the SimCacheModule and we have guarantees that the object will be alive for the duration of the view model.

File: Source/SimCacheModule/Source/Module/SimCacheModule.cpp [hunk 4]
```diff
@@ -34,15 +35,22 @@ bool SimCacheModule::Initialize()
...
+
 	return true;
 }
 
 // -----------------------------------------------------------------------------
 
 void SimCacheModule::Uninitialize()
 {
-	UninitializeSimConnectClient();
+	UninitializeTrackerViewModel();
```
Comment (ollyau) [inline Source/SimCacheModule/Source/Module/SimCacheModule.cpp:51]: I think this is fine, but I'm just wondering if there's a particular reason for the order/sequence here?  Is the idea that you unhook all the things that receive data before you take the data away?

Comment (StevenFrost) [reply Source/SimCacheModule/Source/Module/SimCacheModule.cpp:51]: Uninitializing in the reverse order they were initialized in, so if any of them use another as a dependency the user of the system is destroyed before the system itself

**Review by ollyau — APPROVED** [review]

**Pull request merged**